frontend/src/main/scala/bloop/engine/tasks/Tasks.scala

```diff
@@ -112,53 +128,43 @@ object Tasks {
       testEventHandler: BloopTestSuiteEventHandler,
       runInParallel: Boolean = false,
       mode: RunMode
-  ): Task[State] = {
-    import state.logger
-    implicit val logContext: DebugFilter = DebugFilter.Test
+  ): Task[TestRuns] = {
 
-    var failure = false
     val testTasks = projectsToTest.map { project =>
-      /* Intercept test failures to set the correct error code */
-      val failureHandler = new LoggingEventHandler(state.logger) {
+      val resultsBuilder = List.newBuilder[TestSuiteEvent.Results]
+      val handleAndStore = new LoggingEventHandler(state.logger) {
         override def report(): Unit = testEventHandler.report()
         override def handle(event: TestSuiteEvent): Unit = {
           testEventHandler.handle(event)
           event match {
-            case TestSuiteEvent.Results(_, ev)
-                if ev.exists(e => TestFailedStatus.contains(e.status())) =>
-              failure = true
+            case x: TestSuiteEvent.Results => resultsBuilder += x
             case _ => ()
           }
         }
       }
 
       val cwd = project.workingDirectory
-      TestTask.runTestSuites(
-        state,
-        project,
-        cwd,
-        userTestOptions,
-        testFilter,
-        testClasses,
-        failureHandler,
-        mode
-      )
+      TestTask
+        .runTestSuites(
+          state,
+          project,
+          cwd,
+          userTestOptions,
+          testFilter,
+          testClasses,
+          handleAndStore,
+          mode
+        )
+        .map { exitCode => TestRun(project, exitCode, resultsBuilder.result()) }
     }
 
-    val runAll: List[Task[Int]] => Task[List[Int]] =
-      if (runInParallel) {
-        Task.gather
-      } else {
-        Task.sequence
-      }
+    def runAll[A]: List[Task[A]] => Task[List[A]] =
+      if (runInParallel) Task.gather else Task.sequence
 
-    runAll(testTasks).map { exitCodes =>
+    runAll(testTasks).map { testRuns =>
       // When the test execution is over report no matter what the result is
       testEventHandler.report()
-      logger.debug(s"Test suites failed: $failure")
-      val isOk = !failure && exitCodes.forall(_ == 0)
-      if (isOk) state.mergeStatus(ExitStatus.Ok)
-      else state.copy(status = ExitStatus.TestExecutionError)
+      TestRuns(testRuns)
```

Comment:
I wonder if `testEventHandler` is suitable for handling parallel execution of task, I see mutable collections there and plain Ints :/ but ofc this isn't concern of this PR

Reply:
the test code seems a bit half baked, so I wouldn't be surprised if some fixes are necessary. A related issue I came across is that clients are not (AFAIU) able to request parallel execution at all for now. Perhaps that could be a bloop specific field inside the request data strucure again.